In [1]:
import pandas as pd

### <a href="https://www.kaggle.com/datasets/openfoodfacts/world-food-facts" target="_blank"> $\color{orange}{\text{LINK TO THE DATA}}$</a>

It's a TSV file, so to load the data you can either use pd.read_table('path.tsv') or pd.read_csv('path.tsv', sep='\t')

In [2]:
food_all = pd.read_table('foodfacts.tsv')
food_all.head()

C:\Users\necke\AppData\Local\Temp\ipykernel_2916\823064207.py:1: DtypeWarning: Columns (0,3,5,19,20,24,25,26,27,28,36,37,38,39,48) have mixed types. Specify dtype option on import or set low_memory=False.
  food_all = pd.read_table('foodfacts.tsv')


code                                                url  \
0   3087  http://world-en.openfoodfacts.org/product/0000...   
1   4530  http://world-en.openfoodfacts.org/product/0000...   
2   4559  http://world-en.openfoodfacts.org/product/0000...   
3  16087  http://world-en.openfoodfacts.org/product/0000...   
4  16094  http://world-en.openfoodfacts.org/product/0000...   

                      creator   created_t      created_datetime  \
0  openfoodfacts-contributors  1474103866  2016-09-17T09:17:46Z   
1             usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
2             usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
3             usda-ndb-import  1489055731  2017-03-09T10:35:31Z   
4             usda-ndb-import  1489055653  2017-03-09T10:34:13Z   

  last_modified_t last_modified_datetime                    product_name  \
0      1474103893   2016-09-17T09:18:13Z              Farine de blé noir   
1      1489069957   2017-03-09T14:32:37Z  Banana Chips Sweetened (Whole)   
2      1489069957   2017-03-09T14:32:37Z                         Peanuts   
3      1489055731   2017-03-09T10:35:31Z          Organic Salted Nut Mix   
4      1489055653   2017-03-09T10:34:13Z                 Organic Polenta   

  generic_name quantity  ... fruits-vegetables-nuts_100g  \
0          NaN      1kg  ...                         NaN   
1          NaN      NaN  ...                         NaN   
2          NaN      NaN  ...                         NaN   
3          NaN      NaN  ...                         NaN   
4          NaN      NaN  ...                         NaN   

  fruits-vegetables-nuts-estimate_100g collagen-meat-protein-ratio_100g  \
0                                  NaN                              NaN   
1                                  NaN                              NaN   
2                                  NaN                              NaN   
3                                  NaN                              NaN   
4                                  NaN                              NaN   

  cocoa_100g chlorophyl_100g carbon-footprint_100g nutrition-score-fr_100g  \
0        NaN             NaN                   NaN                     NaN   
1        NaN             NaN                   NaN                    14.0   
2        NaN             NaN                   NaN                     0.0   
3        NaN             NaN                   NaN                    12.0   
4        NaN             NaN                   NaN                     NaN   

  nutrition-score-uk_100g glycemic-index_100g water-hardness_100g  
0                     NaN                 NaN                 NaN  
1                    14.0                 NaN                 NaN  
2                     0.0                 NaN                 NaN  
3                    12.0                 NaN                 NaN  
4                     NaN                 NaN                 NaN  

[5 rows x 163 columns]

There are 163 columns. For simplicity, I choose to pick only a few of them, and then clean the data.

In [13]:
#List of all the columns:
# list(food_all.columns)

#Columns that I choose:
columns = ['product_name', 'fat_100g', 'carbohydrates_100g', 'sugars_100g', 'proteins_100g']

#Assigning it to new DataFrame:
food = food_all[columns].copy(deep=True)

food.head()

product_name  fat_100g  carbohydrates_100g  sugars_100g  \
0              Farine de blé noir       NaN                 NaN          NaN   
1  Banana Chips Sweetened (Whole)     28.57               64.29        14.29   
2                         Peanuts     17.86               60.71        17.86   
3          Organic Salted Nut Mix     57.14               17.86         3.57   
4                 Organic Polenta      1.43               77.14          NaN   

   proteins_100g  
0            NaN  
1           3.57  
2          17.86  
3          17.86  
4           8.57

## DATATYPES AND NULLS

First I check what are the datatypes for each column, and change it if necessary. Then I drop all the redundant rows with null data.

In [14]:
food.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 356027 entries, 0 to 356026
Data columns (total 5 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   product_name        338515 non-null  object 
 1   fat_100g            279497 non-null  float64
 2   carbohydrates_100g  279220 non-null  float64
 3   sugars_100g         279186 non-null  float64
 4   proteins_100g       294161 non-null  float64
dtypes: float64(4), object(1)
memory usage: 13.6+ MB


In [16]:
# It would be better to have a string type in 'product_name' column:
food['product_name'] = food['product_name'].astype('string')
print(food.dtypes['product_name'])

# Rest of the datatypes are fine

string


In [17]:
# I want to get rid of the rows for which there's no product name or any of the macros are NULLs:

print('Number of nulls in product_name column: %d' % food['product_name'].isnull().sum())
food.dropna(subset = ['product_name'], inplace = True)
print('Number of nulls in product_name column: %d' % food['product_name'].isnull().sum())

print('Number of nulls in fat_100g column: %d' % food['fat_100g'].isnull().sum())
food.dropna(subset = ['fat_100g'], inplace = True)
print('Number of nulls in fat_100g column: %d' % food['fat_100g'].isnull().sum())

print('Number of nulls in carbohydrates_100g column: %d' % food['carbohydrates_100g'].isnull().sum())
food.dropna(subset = ['carbohydrates_100g'], inplace = True)
print('Number of nulls in carbohydrates_100g column: %d' % food['carbohydrates_100g'].isnull().sum())

print('Number of nulls in sugars_100g column: %d' % food['sugars_100g'].isnull().sum())
food.dropna(subset = ['sugars_100g'], inplace = True)
print('Number of nulls in sugars_100g column: %d' % food['sugars_100g'].isnull().sum())

print('Number of nulls in proteins_100g column: %d' % food['proteins_100g'].isnull().sum())
food.dropna(subset = ['proteins_100g'], inplace = True)
print('Number of nulls in proteins_100g column: %d' % food['proteins_100g'].isnull().sum())

Number of nulls in product_name column: 17512
Number of nulls in product_name column: 0
Number of nulls in fat_100g column: 62359
Number of nulls in fat_100g column: 0
Number of nulls in carbohydrates_100g column: 1179
Number of nulls in carbohydrates_100g column: 0
Number of nulls in sugars_100g column: 15666
Number of nulls in sugars_100g column: 0
Number of nulls in proteins_100g column: 585
Number of nulls in proteins_100g column: 0


## DUPLICATES

In [19]:
# Let's see if we have duplicates
food[food.duplicated()]

product_name  fat_100g  \
154                                               Peanuts     17.86   
228                              Luxury Christmas Pudding      7.00   
333     The Madelaine Chocolate Company, Solid Milk Ch...     33.33   
367                         Emojeez, Gummies Candy, Fruit      0.00   
394                                 Pb&J Trail Mix, Grape     25.88   
...                                                   ...       ...   
355883                    Chipotle Peppers In Adobo Sauce      0.00   
355884                                        Dal Makhani      5.33   
355955                  Sour Licorice Twists, Sour Cherry      1.25   
355957                                 Taco Seasoning Mix      0.00   
356026                    Sugar Free Drink Mix, Peach Tea      0.00   

        carbohydrates_100g  sugars_100g  proteins_100g  
154                  60.71        17.86          17.86  
228                  54.20        44.00           3.90  
333                  58.33        55.56           8.33  
367                  67.50        40.00           5.00  
394                  51.76        44.71           9.41  
...                    ...          ...            ...  
355883               16.67         6.67           0.00  
355884               13.33         0.00           4.67  
355955               72.50        32.50           2.50  
355957               66.67        16.67          16.67  
356026                0.00         0.00           0.00  

[24618 rows x 5 columns]

In [21]:
# Check for example the first duplicate (Peanuts)
food[food['product_name'] == 'Peanuts'].head()

product_name  fat_100g  carbohydrates_100g  sugars_100g  proteins_100g
2          Peanuts     17.86               60.71        17.86          17.86
154        Peanuts     17.86               60.71        17.86          17.86
4389       Peanuts     46.43               17.86         3.57          25.00
6600       Peanuts     39.29               28.57         3.57          25.00
16219      Peanuts     53.57               17.86         3.57          25.00

Tthere are different values in macros for many products with the same product_name. Therefore, it would be wise to get rid of the duplicates by grouping products by the product_name column and assign for each product its average macros value:

In [24]:
# groups by product_name, so removes duplicates, and puts mean values for each entry which was duplicated
products = pd.DataFrame(food.groupby('product_name').mean())

# sorting by product_name
products.sort_values(by = ['product_name'], inplace = True)

# now you also need to add indexing, because it's gone
products.reset_index(drop=False, inplace=True)

# and round results to decimals = 2
products = products.round(decimals=2)

products

product_name  fat_100g  \
0         Bridelight 3% Les carrés fondants goût Emmental       3.0   
1                                          Salted peanuts      50.0   
2                                     100 % Pur Jus Pomme       0.0   
3                                       18 marrons glacés       0.8   
4                         6 carrés fourrés, saveur amande      14.3   
...                                                   ...       ...   
189415                                                씨그램       0.0   
189416                       자연은 튼튼 (Jayeon-eun Teunteun)       0.0   
189417                                               초고추장       0.0   
189418                                              칠성사이다       0.0   
189419                                                콩두유       6.0   

        carbohydrates_100g  sugars_100g  proteins_100g  
0                     8.50         8.50          13.00  
1                    11.00         5.00          27.00  
2                    10.00         9.70           0.50  
3                    76.00        56.00           0.80  
4                    59.10        34.30           5.50  
...                    ...          ...            ...  
189415                0.00         0.00           0.00  
189416               10.50        10.50           0.00  
189417               47.37        21.05           5.26  
189418               10.80         8.40           0.00  
189419               10.00         7.00           6.00  

[189420 rows x 5 columns]

In [26]:
# revisiting Peanuts (now we have no duplicates):
products[products['product_name']=='Peanuts']

product_name  fat_100g  carbohydrates_100g  sugars_100g  proteins_100g
126285      Peanuts     42.36               29.54        12.69          22.25

## FURTHER INVESTIGATION: LOOKING FOR WRONG DATA

In [27]:
round(products.describe(), 2)

fat_100g  carbohydrates_100g  sugars_100g  proteins_100g
count  189420.00           189420.00    189420.00      189420.00
mean       12.29               33.11        15.91           7.24
std        15.52               28.76        20.77           8.24
min         0.00                0.00        -6.25        -800.00
25%         0.50                6.70         1.35           1.01
50%         6.00               23.33         5.88           5.14
75%        20.00               59.10        24.62          10.01
max       105.00              139.00       104.00         154.25

There are some weird looking results:

* there are negative values of sugars and proteins
* for every product any of the macros shouldn't have value bugger than 100 and they all shoul sum up maximally to 100

Therefore:

* I add column macros_sum = fat + carbs + proteins
* remove data with negative values
* remove data for which macros_sum > 100

In [31]:
# removing negative values

negative_proteins = products[products['proteins_100g'] < 0].index
products.drop(list(negative_proteins), inplace=True)

negative_sugars = products[products['sugars_100g'] < 0].index
products.drop(list(negative_sugars), inplace=True)

# new column: macros_sum
products['macros_sum'] = products['fat_100g']+products['carbohydrates_100g']+products['proteins_100g']

# removing product with macros_sum > 100
too_much_macros = products[products['macros_sum'] > 100].index
products.drop(list(too_much_macros), inplace = True)

round(products.describe(), 2)

fat_100g  carbohydrates_100g  sugars_100g  proteins_100g  macros_sum
count  188020.00           188020.00    188020.00      188020.00   188020.00
mean       12.15               32.86        15.80           7.20       52.22
std        15.40               28.67        20.67           7.97       32.79
min         0.00                0.00         0.00           0.00        0.00
25%         0.44                6.67         1.33           1.00       21.04
50%         5.90               23.00         5.80           5.10       50.00
75%        20.00               58.90        24.24          10.00       86.60
max       100.00              100.00       100.00         100.00      100.00

Now the data is ready for analysis.